hibernate-core/src/main/java/org/hibernate/query/sqm/internal/TypecheckUtil.java

```diff
@@ -329,10 +330,12 @@ public static boolean isTypeAssignable(
 		}
 
 		// entities can be assigned if they belong to the same inheritance hierarchy
-
-		if ( targetType instanceof EntityType<?> targetEntity
-				&& expressionType instanceof EntityType<?> expressionEntity ) {
-			return isEntityTypeAssignable( targetEntity, expressionEntity, bindingContext);
+		if (expressionType instanceof EntityType<?> expressionEntity) {
+			if ( targetType instanceof EntityType<?> targetEntity ) {
+				return isEntityTypeAssignable( targetEntity, expressionEntity, bindingContext );
+			} else if ( targetType instanceof MappedSuperclassDomainType<?> expressionMappedSuperclass ) {
+				return isMappedSuperclassTypeAssignable( expressionMappedSuperclass, expressionEntity, bindingContext );
+			}
```

Comment:
I liked the "old" style better, we should still check the `targetType` first to avoid an unnecessary instanceof on the `expressionType`.